src/coreclr/gc/gc.cpp

```diff
@@ -19469,6 +19472,30 @@ bool gc_heap::init_table_for_region (int gen_number, heap_segment* region)
 }
 #endif //USE_REGIONS
 
+BOOL gc_heap::consumed_enough_gen2_budget_p (dynamic_data* dd_max, uint32_t memory_load)
+{
+    // we computed gen2 budget under some memory load - check if the memory load is significantly higher now
+    uint32_t prev_remaining_memory = 100 - dd_entry_memory_load (dd_max);
+    uint32_t curr_remaining_memory = 100 - memory_load;
+
+    // if the previous memory load was already at the limit, the situation cannot have gotten worse
+    if (prev_remaining_memory <= 0)
+        return FALSE;
+
+    // if the memory load is at the limit now, assume some action is required
+    if (curr_remaining_memory <= 0)
+        return TRUE;
+
+    // check if we have consumed a fraction of the budget that is higher than the factor the situation is worse by now
+    // e.g.: if the memory load had been 90% when we computed the budget, we had 10% free
+    //       if the memory load is 98% now, we have 2% free - so we require at least 20% of the budget to be consumed
+    float budget_fraction_remaining = (float)dd_new_allocation (dd_max) / (float)dd_desired_allocation (dd_max);
+    float budget_fraction_consumed = budget_fraction_remaining > 0.0f ? (1.0f - budget_fraction_remaining) : 1.0f;
+    float memory_fraction_remaining = (float)curr_remaining_memory/(float)prev_remaining_memory;
+
+    return (budget_fraction_consumed >= memory_fraction_remaining);
```

Comment:
I don't understand how this logic works. in most cases we can think that `memory_fraction_remaining` is always 1 because the previous and concurrent memory load are the same (they are integers so on big machines it's very likely it remains the same for a long period of time). and this method would only ever return true if `budget_fraction_consumed` is >= 1.0 which means the budget is fully consumd...

Reply:
Exactly right. The purpose of this method is to adapt to a _changing_ memory load. Adjustment to a _constant_ memory load is done in the budget computation.